src/graph.py

```diff
@@ -752,6 +741,105 @@ def perform_GO_enrichment_on_clusters(nxgraph, clustering):
+    or "specificity", entries indicate robustness or specificity values for each experiment. 
+    Also, node attributes columns are added for plotting purposes.
+
+    Arguments:
+        results (list of tuples): Results of `_grid_randomization` (paramstring, forest, augmented forest)
+        mode (str): Reported values "membership", "robustness", "specificity"
+        top_n (int): Takes the top_n values of the summary dataframe. top_n=-1 sets no threshold
+
+    Returns:
+        pd.DataFrame: Columns correspond to each parameter experiment, indexed by nodes
+    """
+    
+    # Exclude any degenerate results
+    results = {paramstring: graphs for paramstring, graphs in results.items() if graphs["augmented_forest"].number_of_nodes() > 0}
+    
+    if mode == "membership": # Summarize single-run parameter search
+        series = [pd.Series(1, index=graphs["augmented_forest"].nodes(), name=paramstring) for paramstring, graphs in results.items()]
+    elif mode == "robustness": # Summarize randomized robustness
+        series = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["robustness"].rename(paramstring) for paramstring, graphs in results.items()]
+    elif mode == "specificity": # Summarize randomized specificity
+        series = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["specificity"].rename(paramstring) for paramstring, graphs in results.items()]
+    else:
+        logger.info("`mode` must be one of the following: 'membership', 'robustness', or 'specificity'.")
+        return
+    
+    df = pd.concat(series, axis=1).fillna(0)
+    
+    # df can get quite large with many sparse entries, so let's filter for the top_n entries
+    if not top_n: return df
+    
+    if len(df) > top_n: 
+        cutoff = sorted(df.sum(axis=1).tolist(), reverse=True)[top_n]
+        df = df[df.sum(axis=1) > cutoff]
+    
+    return df
+
+
+def get_robust_subgraph_from_randomizations(nxgraph, max_size=400, min_component_size=5): 
+    """
+    Given a graph with robustness attributes, take the top `max_size` robust nodes and
+    prune any "small" components. 
+
+    Arguments:
+        nxgraph (networkx.Graph): Network from randomization experiment
+        max_size (int): Max size of robust network
+
+    Returns: 
+        networkx.Graph: Robust network
+    """
+    
+    # TODO: Potential alternative approach - from entire network, attempt to remove lowest robustness node. 
+    # If removal results in a component of size less than min_size, do not remove. 
+    
+    if nxgraph.number_of_nodes() == 0: 
+        logger.info("Augmented forest is empty.")
+        return nxgraph
+
+    # Get indices of top nodes sorted by high robustness, then low specificity. Don't allow nodes with robustness = 0.
+    node_attributes_df = get_networkx_graph_as_dataframe_of_nodes(nxgraph)
+    node_attributes_df = node_attributes_df[node_attributes_df["robustness"] > 0]
```

Comment:
Are there really nodes with 0 robustness in the graph?

Reply:
Yes, since we merge with specificity experiments. There are a lot.